### PR TITLE
DSDEGP-2390 Require a hard-coded ordering of key fields in ES index.

### PR DESCRIPTION
**Pull request by danxmoran**
### Description

:sadtrombone:

----

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: Ensure that you have added or updated tests
- [ ] **Submitter**: If you're adding/switching libraries or otherwise changing the design, update the [design documentation](https://broadinstitute.atlassian.net/wiki/pages/viewpage.action?pageId=114531509).
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
* Review cycle:
  * Team may comment on PR at will
  * **Reviewer assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to reviewer** for further feedback
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Move the JIRA issue to QA once this checklist is completed

### CODE REVIEW

File: clio-dataaccess-model/src/main/scala/org/broadinstitute/clio/server/dataaccess/elasticsearch/ElasticsearchIndex.scala [hunk 1]
```diff
@@ -24,11 +23,24 @@ import org.broadinstitute.clio.util.model.{DataType, UpsertId}
...
   import clioIndex.implicits._
 
+  private val keyFields = FieldMapper[clioIndex.KeyType].fields.map {
+    case (k, v) => ElasticsearchUtil.toElasticsearchName(k) -> v
+  }
+  private val metadataFields = FieldMapper[clioIndex.MetadataType].fields.map {
+    case (k, v) => ElasticsearchUtil.toElasticsearchName(k) -> v
+  }
+
+  assert(
```
Comment (pshapiro4broad) [inline clio-dataaccess-model/src/main/scala/org/broadinstitute/clio/server/dataaccess/elasticsearch/ElasticsearchIndex.scala:39]: I don't think we should have asserts in our production code. Doing it in a test, for each index, should be sufficient.

File: clio-dataaccess-model/src/main/scala/org/broadinstitute/clio/server/dataaccess/elasticsearch/ElasticsearchIndex.scala [hunk 2]
```diff
@@ -24,11 +23,24 @@ import org.broadinstitute.clio.util.model.{DataType, UpsertId}
 class ElasticsearchIndex[+CI <: ClioIndex](
   val clioIndex: CI,
   final val indexName: String,
+  final val orderedKeyFields: Seq[String],
```
Comment (pshapiro4broad) [inline clio-dataaccess-model/src/main/scala/org/broadinstitute/clio/server/dataaccess/elasticsearch/ElasticsearchIndex.scala:26]: Would be nice to update the params in the doc comment

File: clio-dataaccess-model/src/main/scala/org/broadinstitute/clio/server/dataaccess/elasticsearch/ElasticsearchIndex.scala [hunk 3]
```diff
@@ -24,11 +23,24 @@ import org.broadinstitute.clio.util.model.{DataType, UpsertId}
...
   private[elasticsearch] val fieldMapper: ElasticsearchFieldMapper,
   private[elasticsearch] val defaultFields: Json = Json.obj()
 ) extends ModelAutoDerivation {
   import clioIndex.implicits._
 
+  private val keyFields = FieldMapper[clioIndex.KeyType].fields.map {
+    case (k, v) => ElasticsearchUtil.toElasticsearchName(k) -> v
```
Comment (pshapiro4broad) [inline clio-dataaccess-model/src/main/scala/org/broadinstitute/clio/server/dataaccess/elasticsearch/ElasticsearchIndex.scala:33]: Why is this needed now that we have `orderedKeyFields`?

File: clio-dataaccess-model/src/main/scala/org/broadinstitute/clio/server/dataaccess/elasticsearch/ElasticsearchIndex.scala (hunk 1 above)
Comment (danxmoran) [reply clio-dataaccess-model/src/main/scala/org/broadinstitute/clio/server/dataaccess/elasticsearch/ElasticsearchIndex.scala:39]: I worry about having no guards here, where before we at least had one source of truth (the type declaration). Tests will work until we add the next index and forget about to add the very-important-unit-test.

If I changed it to `require` instead of `assert`, would that be better? It throws an `IllegalArgumentException` instead of an `AssertionError`.

File: clio-dataaccess-model/src/main/scala/org/broadinstitute/clio/server/dataaccess/elasticsearch/ElasticsearchIndex.scala (hunk 3 above)
Comment (danxmoran) [reply clio-dataaccess-model/src/main/scala/org/broadinstitute/clio/server/dataaccess/elasticsearch/ElasticsearchIndex.scala:33]: It's needed to verify that `orderedKeyFields` contains all the required pieces. Will remove if we end up dropping that check from the main code.

**Review by tbl3rd — APPROVED** [review]
